# Exploration & Wrangling Data of the Crimes in Chicago Dataset 

* The data is extracted from: https://www.kaggle.com/currie32/crimes-in-chicago/data

## Importing Basic Packages 
These packages help us rearrange our data into a dataframe format easily 

In [1]:
import numpy as np  
import pandas as pd 
import math

## Import 4 Datasets 

These 4 datasets are the ones downloaded from Kaggle. They should be in the crimesInChicagoData folder. 

In [2]:
df1 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2001_to_2004.csv", error_bad_lines = False)

b'Skipping line 1513591: expected 23 fields, saw 24\n'
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (17,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df2 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2005_to_2007.csv", error_bad_lines = False)

b'Skipping line 533719: expected 23 fields, saw 24\n'


In [4]:
df3 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2008_to_2011.csv", error_bad_lines = False)

b'Skipping line 1149094: expected 23 fields, saw 41\n'


In [5]:
df4 = pd.read_csv("../../crimesInChicagoData/Chicago_Crimes_2012_to_2017.csv", error_bad_lines = False)

## Analyze  Missing Data 
Seeing how much data is missing per column of each of the 4 dataframes to get an understanding of how much missing data we have.

In [6]:
'''Takes in a dataframe and returns a dataframe with percentages of missing data of each column
'''
def missingDataSummary(df):
    dfInitial = pd.DataFrame(index = [df.name], columns =df.columns)
    for column in df.columns:
        #print("Column name: " + column )
        numNas = df[column].isnull().sum()
        numObservations = len(df[column])
        missingDataPercentage = numNas/numObservations *100
        dfInitial[column][0] = missingDataPercentage
        #print(str(missingDataPercentage) + "% missing") 
    return dfInitial 

In [7]:
#set the names of the dataframes for convenience 
df1.name = 'df1'
df2.name = 'df2'
df3.name = 'df3'
df4.name = 'df4'

In [8]:
#retrieve info on missing data for each data frame 
df1_missingSummary = missingDataSummary(df1)
df2_missingSummary  = missingDataSummary(df2)
df3_missingSummary = missingDataSummary(df3)
df4_missingSummary = missingDataSummary(df4)

#concatenate missingData summary info into one
df_missingSummaries = pd.concat([df1_missingSummary, df2_missingSummary, df3_missingSummary, df4_missingSummary], axis = 0)

#display all the missing data summary
pd.set_option('display.max_columns', None)
df_missingSummaries

Unnamed: 0 ID  Case Number Date Block IUCR Primary Type Description  \
df1          0  0            0    0     0    0            0           0   
df2          0  0            0    0     0    0            0           0   
df3          0  0  0.000223155    0     0    0            0           0   
df4          0  0  6.86477e-05    0     0    0            0           0   

    Location Description Arrest Domestic Beat     District         Ward  \
df1          0.000831811      0        0    0  0.000103976      36.3986   
df2           0.00133523      0        0    0  0.000267045  0.000801135   
df3             0.010823      0        0    0   0.00308698   0.00234313   
df4             0.113818      0        0    0  6.86477e-05  0.000961067   

    Community Area FBI Code X Coordinate Y Coordinate Year Updated On  \
df1        36.4046        0      1.59557      1.59557    0          0   
df2      0.0186397        0     0.488906     0.488906    0          0   
df3      0.0541152        0      1.06538      1.06538    0          0   
df4     0.00274591        0      2.54566      2.54566    0          0   

     Latitude Longitude  Location  
df1   1.59557   1.59562   1.59562  
df2  0.488906  0.488906  0.488906  
df3   1.06538   1.06538   1.06538  
df4   2.54566   2.54566   2.54566

## Combine Four Datasets into One
Concatenate 4 data sets to make one total data set form years range 2001-2017

In [9]:
dfTotal = pd.concat([df1, df2, df3, df4], axis = 0)

Checking if the concatenation indeed happened:

In [10]:
print("Are the number of columns maintained?")
print(len(df1.columns) == len(df2.columns) ==  len(df3.columns) == len(df4.columns) == len(dfTotal.columns))

print("Are the number of observations of all data sets equal to the dfTotal")
print(len(df1['ID']) + len(df2['ID']) + len(df3['ID']) + len(df4['ID']) == len(dfTotal['ID']))

dfTotal.columns

Are the number of columns maintained?
True
Are the number of observations of all data sets equal to the dfTotal
True


Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [11]:
dfTotal.head()

Unnamed: 0       ID Case Number                    Date  \
0         879  4786321    HM399414  01/01/2004 12:01:00 AM   
1        2544  4676906    HM278933  03/01/2003 12:00:00 AM   
2        2919  4789749    HM402220  06/20/2004 11:00:00 AM   
3        2927  4789765    HM402058  12/30/2004 08:00:00 PM   
4        3302  4677901    HM275615  05/01/2003 01:00:00 AM   

                 Block  IUCR                Primary Type  \
0    082XX S COLES AVE  0840                       THEFT   
1      004XX W 42ND PL  2825               OTHER OFFENSE   
2  025XX N KIMBALL AVE  1752  OFFENSE INVOLVING CHILDREN   
3   045XX W MONTANA ST  0840                       THEFT   
4   111XX S NORMAL AVE  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   Beat  District  Ward  Community Area FBI Code  X Coordinate Y Coordinate  \
0   424       4.0   7.0            46.0       06           NaN          NaN   
1   935       9.0  11.0            61.0       26     1173974.0  1.87676e+06   
2  1413      14.0  35.0            22.0       20           NaN          NaN   
3  2521      25.0  31.0            20.0       06           NaN          NaN   
4  2233      22.0  34.0            49.0       06     1174948.0  1.83105e+06   

     Year              Updated On Latitude  Longitude  \
0  2004.0  08/17/2015 03:03:40 PM      NaN        NaN   
1  2003.0  04/15/2016 08:55:02 AM  41.8172 -87.637328   
2  2004.0  08/17/2015 03:03:40 PM      NaN        NaN   
3  2004.0  08/17/2015 03:03:40 PM      NaN        NaN   
4  2003.0  04/15/2016 08:55:02 AM  41.6918 -87.635116   

                        Location  
0                            NaN  
1  (41.817229156, -87.637328162)  
2                            NaN  
3                            NaN  
4  (41.691784636, -87.635115968)

## Conversion of Time Data 
* We converted the 'Date' data type to a Time Series for easy sorting in pandas df. This sorting mechanism allows us to explore the data easily. 
* We also converted the format of the time. We got rid of AM & PM designations and made the hours from a 12-hour clock to a 24-hour clock. This was in part so that we could look at the data by hours more easily. 

In [17]:
dfTotal.sort_values(by='Year').head(20)

Unnamed: 0       ID Case Number                    Date  \
1602847     3629582  1423259     G139165  03/10/2001 11:30:00 PM   
386540      3926865  1797113     G623639  10/17/2001 06:00:00 AM   
382057      3922382  1791530     G615404  10/13/2001 01:30:00 AM   
382058      3922383  1791531     G614237  10/12/2001 10:03:49 PM   
382059      3922384  1791532     G614502  10/13/2001 12:55:00 AM   
382060      3922385  1791533     G615483  10/10/2001 10:00:00 AM   
382061      3922386  1791534     G615214  10/05/2001 12:00:00 PM   
382062      3922387  1791535     G614706  10/13/2001 03:10:00 AM   
382063      3922388  1791536     G614719  10/13/2001 02:55:00 AM   
382064      3922389  1791537     G614541  10/13/2001 01:04:29 AM   
382065      3922390  1791538     G614847  10/13/2001 05:48:04 AM   
382066      3922391  1791539     G615431  10/04/2001 01:30:00 PM   
382067      3922392  1791540     G615507  10/12/2001 02:30:00 AM   
382068      3922393  1791542     G615510  10/13/2001 12:00:00 AM   
382069      3922394  1791544     G615478  10/12/2001 10:00:00 PM   
382056      3922381  1791529     G615430  10/13/2001 01:00:00 PM   
382055      3922380  1791528     G615173  10/12/2001 05:00:00 PM   
382054      3922379  1791527     G615446  10/13/2001 02:00:00 AM   
382053      3922378  1791526     G615450  10/12/2001 08:30:00 PM   
382052      3922377  1791525     G614137  10/12/2001 09:45:00 PM   

                         Block  IUCR        Primary Type  \
1602847     035XX S FEDERAL ST  1340     CRIMINAL DAMAGE   
386540     007XX S COLUMBUS DR  1320     CRIMINAL DAMAGE   
382057     013XX W MONTROSE AV  0820               THEFT   
382058        001XX N LOTUS AV  0460             BATTERY   
382059         007XX N PINE AV  0460             BATTERY   
382060      013XX W BELMONT AV  0820               THEFT   
382061       017XX W NELSON ST  1310     CRIMINAL DAMAGE   
382062     065XX S SANGAMON ST  0820               THEFT   
382063     002XX W GARFIELD BL  0420             BATTERY   
382064     059XX W DIVISION ST  0460             BATTERY   
382065          059XX S MAY ST  1220  DECEPTIVE PRACTICE   
382066     006XX N MICHIGAN AV  2820       OTHER OFFENSE   
382067   055XX W WASHINGTON BL  1310     CRIMINAL DAMAGE   
382068   038XX N CALIFORNIA AV  1320     CRIMINAL DAMAGE   
382069        041XX N DAMEN AV  0820               THEFT   
382056       008XX W ROSCOE ST  0820               THEFT   
382055     027XX N LAKEVIEW AV  0810               THEFT   
382054        054XX W NORTH AV  0820               THEFT   
382053       011XX S LAFLIN ST  0810               THEFT   
382052   055XX W WASHINGTON BL  041A             BATTERY   

                            Description          Location Description  Arrest  \
1602847               TO STATE SUP PROP       CHA PARKING LOT/GROUNDS    True   
386540                       TO VEHICLE                        STREET   False   
382057                   $500 AND UNDER                        STREET   False   
382058                           SIMPLE                      SIDEWALK   False   
382059                           SIMPLE                         ALLEY   False   
382060                   $500 AND UNDER  COMMERCIAL / BUSINESS OFFICE   False   
382061                      TO PROPERTY              RESIDENCE-GARAGE   False   
382062                   $500 AND UNDER                     RESIDENCE   False   
382063   AGGRAVATED:KNIFE/CUTTING INSTR                        STREET   False   
382064                           SIMPLE                      SIDEWALK   False   
382065       THEFT OF LOST/MISLAID PROP                      SIDEWALK    True   
382066                 TELEPHONE THREAT            SMALL RETAIL STORE   False   
382067                      TO PROPERTY                     RESIDENCE   False   
382068                       TO VEHICLE                        STREET   False   
382069                   $500 AND UNDER                        STREET   False   
382056                 

In [13]:
def changeToDateTime(df, columnName, timeFormat): 
    df[columnName] = pd.to_datetime(df[columnName], format = timeFormat)

In [14]:
dateFormat = '%m/%d/%Y %I:%M:%S %p' 
yearFormat = '%Y.0'

In [18]:
changeToDateTime(dfTotal, 'Date', dateFormat)

In [19]:
dfTotal.head()

Unnamed: 0       ID Case Number                Date                Block  \
0         879  4786321    HM399414 2004-01-01 00:01:00    082XX S COLES AVE   
1        2544  4676906    HM278933 2003-03-01 00:00:00      004XX W 42ND PL   
2        2919  4789749    HM402220 2004-06-20 11:00:00  025XX N KIMBALL AVE   
3        2927  4789765    HM402058 2004-12-30 20:00:00   045XX W MONTANA ST   
4        3302  4677901    HM275615 2003-05-01 01:00:00   111XX S NORMAL AVE   

   IUCR                Primary Type                     Description  \
0  0840                       THEFT   FINANCIAL ID THEFT: OVER $300   
1  2825               OTHER OFFENSE         HARASSMENT BY TELEPHONE   
2  1752  OFFENSE INVOLVING CHILDREN   AGG CRIM SEX ABUSE FAM MEMBER   
3  0840                       THEFT   FINANCIAL ID THEFT: OVER $300   
4  0841                       THEFT  FINANCIAL ID THEFT:$300 &UNDER   

  Location Description  Arrest  Domestic  Beat  District  Ward  \
0            RESIDENCE   False     False   424       4.0   7.0   
1            RESIDENCE   False      True   935       9.0  11.0   
2            RESIDENCE   False     False  1413      14.0  35.0   
3                OTHER   False     False  2521      25.0  31.0   
4            RESIDENCE   False     False  2233      22.0  34.0   

   Community Area FBI Code  X Coordinate Y Coordinate    Year  \
0            46.0       06           NaN          NaN  2004.0   
1            61.0       26     1173974.0  1.87676e+06  2003.0   
2            22.0       20           NaN          NaN  2004.0   
3            20.0       06           NaN          NaN  2004.0   
4            49.0       06     1174948.0  1.83105e+06  2003.0   

               Updated On Latitude  Longitude                       Location  
0  08/17/2015 03:03:40 PM      NaN        NaN                            NaN  
1  04/15/2016 08:55:02 AM  41.8172 -87.637328  (41.817229156, -87.637328162)  
2  08/17/2015 03:03:40 PM      NaN        NaN                            NaN  
3  08/17/2015 03:03:40 PM      NaN        NaN                            NaN  
4  04/15/2016 08:55:02 AM  41.6918 -87.635116  (41.691784636, -87.635115968)

In [16]:
#changeToDateTime(df1, 'Year', yearFormat)

In [17]:
#df2003 = df1.loc[df1['Date'].dt.year == 2003] #allows easy slicing of data by time

## Exploring Null Values in Year 2001 'Ward' 
* From the "Analyze Missing Data" section, we found out that df1 had a lot of missing data for the 'Ward' column. We decided to explore it and understand if it was a phenomenon that occurred only for a particular year. 
* We found out that 2001 accounted for563443/7000132 or ~80% of df1's missing 'Ward' data
* We hypothesize that this may have been the case because 'Ward' data may have been put in only after the mid 2001s (we do not know if this true) 

This is the observations in df1 that has all the missing 'Ward' info:

In [18]:
df1WardNull = df1.loc[df1['Ward'].isnull()] 

There are this many missing 'Ward' data in df1's observations:

In [19]:
len(df1WardNull)

700132

These are how many 'Ward' observations are missing in the year 2001: 

In [20]:
ward2001Nan = df1.loc[(df1['Ward'].isnull()) &( df1['Date'].dt.year == 2001)]

len(ward2001Nan)

563443

So 2001 accounts for this percentage of df1's missing observations in 'Ward': 

In [21]:
str(563443/700132*100) + "%" 

'80.47668154005245%'

In [22]:
df2001 = len(df1.loc[df1['Date'].dt.year == 2001])

## Exploration of Unique Labels
* We wanted to see what labels were used for each category/column to better understand the data we were working with 

In [20]:
'''Takes in a dataframe and returns a dataframe with the corresponding
unique labels made in each column
'''
def uniqueLabels(df):
    dfInitial = pd.DataFrame(index = [0], columns =df.columns)
    for column in df.columns:
        uniqueCounts = df[column].unique()
        numObservations = len(df[column])
        dfInitial[column][0] = uniqueCounts
    return dfInitial 

In [21]:
dfTotal.name = 'dfTotal'

In [22]:
dftotal_unique = uniqueLabels(dfTotal)

### (1) The Unique Labels for each Column
This demonstrates and summarizes what kind of labels were used per column: 

In [23]:
dftotal_unique

Unnamed: 0  \
0  [879, 2544, 2919, 2927, 3302, 3633, 3756, 4502...   

                                                  ID  \
0  [4786321, 4676906, 4789749, 4789765, 4677901, ...   

                                         Case Number  \
0  [HM399414, HM278933, HM402220, HM402058, HM275...   

                                                Date  \
0  [2004-01-01T00:01:00.000000000, 2003-03-01T00:...   

                                               Block  \
0  [082XX S COLES AVE, 004XX W 42ND PL, 025XX N K...   

                                                IUCR  \
0  [0840, 2825, 1752, 0841, 0266, 5007, 0890, 175...   

                                        Primary Type  \
0  [THEFT, OTHER OFFENSE, OFFENSE INVOLVING CHILD...   

                                         Description  \
0  [FINANCIAL ID THEFT: OVER $300, HARASSMENT BY ...   

                                Location Description         Arrest  \
0  [RESIDENCE, OTHER, APARTMENT, RESIDENCE PORCH/...  [False, True]   

        Domestic                                               Beat  \
0  [False, True]  [424, 935, 1413, 2521, 2233, 1011, 531, 2222, ...   

                                            District  \
0  [4.0, 9.0, 14.0, 25.0, 22.0, 10.0, 5.0, 18.0, ...   

                                                Ward  \
0  [7.0, 11.0, 35.0, 31.0, 34.0, 24.0, 9.0, 21.0,...   

                                      Community Area  \
0  [46.0, 61.0, 22.0, 20.0, 49.0, 29.0, 50.0, 73....   

                                            FBI Code  \
0  [06, 26, 20, 02, 07, 17, 11, 10, 08B, 05, 15, ...   

                                        X Coordinate  \
0  [nan, 1173974.0, 1174948.0, 1182247.0, 1169911...   

                                        Y Coordinate  \
0  [nan, 1876757.0, 1831051.0, 1829375.0, 1844832...   

                                                Year  \
0  [2004.0, 2003.0, 2001.0, 2002.0, 41.789832136,...   

                                          Updated On  \
0  [08/17/2015 03:03:40 PM, 04/15/2016 08:55:02 A...   

                                            Latitude  \
0  [nan, 41.817229155999996, 41.691784636, 41.687...   

                                           Longitude  \
0  [nan, -87.637328162, -87.635115968, -87.608445...   

                                            Location  
0  [nan, (41.817229156, -87.637328162), (41.69178...

In [24]:
df_missingSummaries

Unnamed: 0 ID  Case Number Date Block IUCR Primary Type Description  \
df1          0  0            0    0     0    0            0           0   
df2          0  0            0    0     0    0            0           0   
df3          0  0  0.000223155    0     0    0            0           0   
df4          0  0  6.86477e-05    0     0    0            0           0   

    Location Description Arrest Domestic Beat     District         Ward  \
df1          0.000831811      0        0    0  0.000103976      36.3986   
df2           0.00133523      0        0    0  0.000267045  0.000801135   
df3             0.010823      0        0    0   0.00308698   0.00234313   
df4             0.113818      0        0    0  6.86477e-05  0.000961067   

    Community Area FBI Code X Coordinate Y Coordinate Year Updated On  \
df1        36.4046        0      1.59557      1.59557    0          0   
df2      0.0186397        0     0.488906     0.488906    0          0   
df3      0.0541152        0      1.06538      1.06538    0          0   
df4     0.00274591        0      2.54566      2.54566    0          0   

     Latitude Longitude  Location  
df1   1.59557   1.59562   1.59562  
df2  0.488906  0.488906  0.488906  
df3   1.06538   1.06538   1.06538  
df4   2.54566   2.54566   2.54566

### (2) Printing Out the Number of Unique Labels Per Column

In [25]:
def printUniqueSummaryValues (df): 
    for column in df.columns: 
        print(column + " has these many unique values: ")
        print(len(df[column][0]))

In [26]:
printUniqueSummaryValues(dftotal_unique)

Unnamed: 0 has these many unique values: 
6170812
ID has these many unique values: 
6170812
Case Number has these many unique values: 
6170473
Date has these many unique values: 
2451622
Block has these many unique values: 
58776
IUCR has these many unique values: 
398
Primary Type has these many unique values: 
35
Description has these many unique values: 
376
Location Description has these many unique values: 
173
Arrest has these many unique values: 
2
Domestic has these many unique values: 
2
Beat has these many unique values: 
304
District has these many unique values: 
27
Ward has these many unique values: 
51
Community Area has these many unique values: 
79
FBI Code has these many unique values: 
26
X Coordinate has these many unique values: 
78276
Y Coordinate has these many unique values: 
152136
Year has these many unique values: 
18
Updated On has these many unique values: 
1310
Latitude has these many unique values: 
864639
Longitude has these many unique values: 
838433
Lo

## Dropping Columns 
From our exploration of labels and Ward info, we were able to understand some of the properties of the categories and realized that some of the columns were too repetitive, too detailed, or unnecessary for answering our question. In addition, we also had problems working and creating a model because our dataset was big and decided to shrink its size. Thus, we decided to drop the following information: 

### Dropping Columns: Part I 

* Unnamed because it was an extraneous column 
* Ward because the dataset had a lot of missing data for df1 and we already had other categories indicating district or location. Hence, we deemed the column repetitive and unnecessary
* Longitude & Latitude because Location includes both coordinates' information  
* Updated On because we are not interested in this data. It does not help answer our question
* Case Number in this case because we already have an ID that identifies each observation uniquely  


In [27]:
dfTotal.columns

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')

In [28]:
delInitialCols = ['Unnamed: 0', 'Case Number', 'Ward', 'X Coordinate', 'Y Coordinate', 'Updated On']

In [29]:
dfTotal = dfTotal.drop(delInitialCols, axis = 1) #actual deletion 

In [30]:
dfTotal.head()

ID                Date                Block  IUCR  \
0  4786321 2004-01-01 00:01:00    082XX S COLES AVE  0840   
1  4676906 2003-03-01 00:00:00      004XX W 42ND PL  2825   
2  4789749 2004-06-20 11:00:00  025XX N KIMBALL AVE  1752   
3  4789765 2004-12-30 20:00:00   045XX W MONTANA ST  0840   
4  4677901 2003-05-01 01:00:00   111XX S NORMAL AVE  0841   

                 Primary Type                     Description  \
0                       THEFT   FINANCIAL ID THEFT: OVER $300   
1               OTHER OFFENSE         HARASSMENT BY TELEPHONE   
2  OFFENSE INVOLVING CHILDREN   AGG CRIM SEX ABUSE FAM MEMBER   
3                       THEFT   FINANCIAL ID THEFT: OVER $300   
4                       THEFT  FINANCIAL ID THEFT:$300 &UNDER   

  Location Description  Arrest  Domestic  Beat  District  Community Area  \
0            RESIDENCE   False     False   424       4.0            46.0   
1            RESIDENCE   False      True   935       9.0            61.0   
2            RESIDENCE   False     False  1413      14.0            22.0   
3                OTHER   False     False  2521      25.0            20.0   
4            RESIDENCE   False     False  2233      22.0            49.0   

  FBI Code    Year Latitude  Longitude                       Location  
0       06  2004.0      NaN        NaN                            NaN  
1       26  2003.0  41.8172 -87.637328  (41.817229156, -87.637328162)  
2       20  2004.0      NaN        NaN                            NaN  
3       06  2004.0      NaN        NaN                            NaN  
4       06  2003.0  41.6918 -87.635116  (41.691784636, -87.635115968)

### Dropping Columns: Part II 
We delete:
* Community Area because a groupings by region, Beat and District, already exist 
* Block because it contains addresses which are too fine grain details that we do not need 
* Location because the Latitude Longitue already includes that info

In [31]:
delInitialCols2 = ['Community Area', 'Block', 'Location']

In [32]:
dfTotal = dfTotal.drop(delInitialCols2, axis = 1) #actual deletion 

In [33]:
dfTotal.head()

ID                Date  IUCR                Primary Type  \
0  4786321 2004-01-01 00:01:00  0840                       THEFT   
1  4676906 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2  4789749 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3  4789765 2004-12-30 20:00:00  0840                       THEFT   
4  4677901 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   Beat  District FBI Code    Year Latitude  Longitude  
0   424       4.0       06  2004.0      NaN        NaN  
1   935       9.0       26  2003.0  41.8172 -87.637328  
2  1413      14.0       20  2004.0      NaN        NaN  
3  2521      25.0       06  2004.0      NaN        NaN  
4  2233      22.0       06  2003.0  41.6918 -87.635116

In [34]:
def printUniqueValues (df): 
    for column in df.columns: 
        print(column + " has these many unique values: ")
        print(len(df[column].unique()))

In [35]:
printUniqueValues(dfTotal)

ID has these many unique values: 
6170812
Date has these many unique values: 
2451622
IUCR has these many unique values: 
398
Primary Type has these many unique values: 
35
Description has these many unique values: 
376
Location Description has these many unique values: 
173
Arrest has these many unique values: 
2
Domestic has these many unique values: 
2
Beat has these many unique values: 
304
District has these many unique values: 
27
FBI Code has these many unique values: 
26
Year has these many unique values: 
18
Latitude has these many unique values: 
864639
Longitude has these many unique values: 
838433


### Dropping Columns: Part III 
* We decided to drop Beat.

While Beat has 0 missing data values, it has 304 different categories while District has 27 and has missing categories. We believed groupings of districts did not need to be so fine grain, particularly when we already have information with Location. 

In [36]:
dfTotal = dfTotal.drop(['Beat'], axis = 1) #actual deletion 

In [37]:
len(dfTotal)

7941282

In [38]:
dfTotal.head()

ID                Date  IUCR                Primary Type  \
0  4786321 2004-01-01 00:01:00  0840                       THEFT   
1  4676906 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2  4789749 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3  4789765 2004-12-30 20:00:00  0840                       THEFT   
4  4677901 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   District FBI Code    Year Latitude  Longitude  
0       4.0       06  2004.0      NaN        NaN  
1       9.0       26  2003.0  41.8172 -87.637328  
2      14.0       20  2004.0      NaN        NaN  
3      25.0       06  2004.0      NaN        NaN  
4      22.0       06  2003.0  41.6918 -87.635116

In [39]:
dfTotal = dfTotal.reset_index() #to realign the indices from the concatenation of four dataframesfrom before

In [40]:
newTotalDf = dfTotal
newTotalDf.head()

index       ID                Date  IUCR                Primary Type  \
0      0  4786321 2004-01-01 00:01:00  0840                       THEFT   
1      1  4676906 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2      2  4789749 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3      3  4789765 2004-12-30 20:00:00  0840                       THEFT   
4      4  4677901 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   District FBI Code    Year Latitude  Longitude  
0       4.0       06  2004.0      NaN        NaN  
1       9.0       26  2003.0  41.8172 -87.637328  
2      14.0       20  2004.0      NaN        NaN  
3      25.0       06  2004.0      NaN        NaN  
4      22.0       06  2003.0  41.6918 -87.635116

### Dropping Columns: Part IV 
* We decided to drop ID code because the information was too fine grain
* We decided to drop FBI Code because we already had information on the type of crime with IUCR
* We dropped index bc it was unncessary 

In [88]:
dataset = newTotalDf.drop(['index', 'ID', 'FBI Code'], axis =1)

In [89]:
dataset.head()

Date  IUCR                Primary Type  \
0 2004-01-01 00:01:00  0840                       THEFT   
1 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3 2004-12-30 20:00:00  0840                       THEFT   
4 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

   District    Year Latitude  Longitude  
0       4.0  2004.0      NaN        NaN  
1       9.0  2003.0  41.8172 -87.637328  
2      14.0  2004.0      NaN        NaN  
3      25.0  2004.0      NaN        NaN  
4      22.0  2003.0  41.6918 -87.635116

## Dropping Missing Observations 

We found out that approxiamtely 1.3% of our observations were missing latitude and longitude data, and ~ 0.02%  of Location Description and ~ 0.001% of District.  

* We decided to fill in 'Location Description's missing values as 'UNKNOWN' because that would identify a type of category for it and also prevent us from losing data. 
* We decided to fill in missing 'Latitude' and 'Longitude' data with 0's so that if we map them out we know for a fact that they are missing values. We also wanted to retain the other data, even if we didn't have the coordinate info. In addition, our future clusterings or models do not rely on the latitude and longitude data. The location information is only for the visualization/mapping. 
* We decided to fill in the district as an 'UNKNOWN' as well so we could retain the reset of the info, but also not incorrectly categorize the group of District the observation belongs to incorreclty because of our lack of expertise in the field.

In [54]:
dataset.name = 'dataset'

In [56]:
missingDatasetValues = missingDataSummary(dataset)

In [57]:
missingDatasetValues

Date IUCR Primary Type Description Location Description Arrest  \
dataset    0    0            0           0            0.0250589      0   

        Domestic    District Year Latitude Longitude  
dataset        0  0.00114591    0  1.32942   1.32943

In [74]:
nullLoc = dataset.loc[dataset['Location Description'].isnull() == True]
nullLoc

Date  IUCR            Primary Type  \
95688   2001-03-18 21:37:10  0460                 BATTERY   
101322  2001-03-22 00:30:00  1200      DECEPTIVE PRACTICE   
490135  2004-08-01 00:05:00  1153      DECEPTIVE PRACTICE   
518707  2002-01-26 16:25:00  2820           OTHER OFFENSE   
731355  2002-07-11 03:00:00  0930     MOTOR VEHICLE THEFT   
770238  2002-08-07 04:00:00  0261     CRIM SEXUAL ASSAULT   
922562  2002-11-23 21:00:00  0890                   THEFT   
1218488 2003-07-10 11:00:00  1150      DECEPTIVE PRACTICE   
1304465 2003-09-16 15:40:00  0560                 ASSAULT   
1469541 2004-01-29 00:01:00  1310         CRIMINAL DAMAGE   
1635389 2004-05-31 00:00:00  0890                   THEFT   
1775989 2004-09-12 20:00:00  0610                BURGLARY   
1776323 2004-09-14 00:45:00  0630                BURGLARY   
1806575 2004-10-02 22:10:16  143A       WEAPONS VIOLATION   
1923398 2004-01-01 14:30:00  1153      DECEPTIVE PRACTICE   
1923400 2004-08-17 10:05:00  1153      DECEPTIVE PRACTICE   
2154372 2006-10-10 06:00:00  1305         CRIMINAL DAMAGE   
2184414 2006-11-02 21:00:00  0820                   THEFT   
2242391 2006-12-16 20:00:00  1320         CRIMINAL DAMAGE   
2295942 2006-11-28 21:48:05  1811               NARCOTICS   
2364230 2007-04-04 21:38:57  1350       CRIMINAL TRESPASS   
2688088 2006-10-10 06:00:00  1305         CRIMINAL DAMAGE   
2718130 2006-11-02 21:00:00  0820                   THEFT   
2776107 2006-12-16 20:00:00  1320         CRIMINAL DAMAGE   
2829658 2006-11-28 21:48:05  1811               NARCOTICS   
2897946 2007-04-04 21:38:57  1350       CRIMINAL TRESPASS   
3079349 2007-08-23 08:00:00  0820                   THEFT   
3213186 2007-12-10 21:05:00  2850  PUBLIC PEACE VIOLATION   
3243610 2005-05-28 19:00:00  1153      DECEPTIVE PRACTICE   
3247053 2007-10-19 23:40:00  1153      DECEPTIVE PRACTICE   
...                     ...   ...                     ...   
7934572 2017-01-09 12:20:00  1153      DECEPTIVE PRACTICE   
7934630 2016-10-06 18:00:00  1153      DECEPTIVE PRACTICE   
7934631 2017-01-05 18:55:00  1153      DECEPTIVE PRACTICE   
7934632 2016-12-15 19:10:00  1153      DECEPTIVE PRACTICE   
7934961 2017-01-10 13:10:00  1154      DECEPTIVE PRACTICE   
7935212 2017-01-10 00:00:00  1153      DECEPTIVE PRACTICE   
7935325 2017-01-02 15:10:00  1154      DECEPTIVE PRACTICE   
7935644 2017-01-11 14:00:00  1154      DECEPTIVE PRACTICE   
7936026 2016-10-14 07:00:00  1153      DECEPTIVE PRACTICE   
7936027 2016-08-01 11:40:00  1153      DECEPTIVE PRACTICE   
7936081 2016-12-26 13:00:00  1153      DECEPTIVE PRACTICE   
7936347 2017-01-12 11:50:00  1153      DECEPTIVE PRACTICE   
7936737 2016-11-18 18:40:00  1153      DECEPTIVE PRACTICE   
7937378 2017-01-13 14:18:00  1153      DECEPTIVE PRACTICE   
7937412 2017-01-13 16:00:00  1153      DECEPTIVE PRACTICE   
7938327 2017-01-07 07:00:00  1153      DECEPTIVE PRACTICE   
7938329 2017-01-14 15:05:00  1153      DECEPTIVE PRACTICE   
7939659 2017-01-03 17:45:00  1153      DECEPTIVE PRACTICE   
7939660 2016-08-04 16:15:00  1153      DECEPTIVE PRACTICE   
7939661 2017-01-02 17:45:00  1153      DECEPTIVE PRACTICE   
7940426 2015-10-23 00:00:00  1153      DECEPTIVE PRACTICE   
7940427 2017-01-10 12:50:00  1153      DECEPTIVE PRACTICE   
7940428 2017-01-17 08:00:00  1153      DECEPTIVE PRACTICE   
7940770 2017-01-18 13:50:00  1154      DECEPTIVE PRACTICE   
7941079 2017-01-18 09:25:00  1154      DECEPTIVE PRACTICE   
7941178 2017-01-09 16:15:00  1153      DECEPTIVE PRACTICE   
7941180 2017-01-10 22:35:00  1153      DECEPTIVE PRACTICE   
7941181 2016-12-28 16:20:00  1153      DECEPTIVE PRACTICE   
7941182 2017-01-03 09:00:00  1153      DECEPTIVE PRACTICE   
7941183 2015-04-15 08:35:00  1153      DECEPTIVE PRACTICE   

                                     Description Location Description  Arrest  \
95688                                     SIMPLE                  NaN   False   
101322             STOLEN PROP: BUY/RECEIVE/POS.                  NaN    True 

Number of missing location description observations: 

In [75]:
len(nullLoc)

1990

In [92]:
dataset['Location Description'] = dataset['Location Description'].fillna('UNKNOWN')

In [93]:
dataset['Location Description'].isnull().sum()

0

In [71]:
nullDistrict = dataset.loc[dataset['District'].isnull() == True]
nullDistrict

Date  IUCR         Primary Type  \
1802297 2004-06-26 10:00:00  1150   DECEPTIVE PRACTICE   
1817066 2004-10-14 15:41:00  0610             BURGLARY   
1969394 2006-05-10 19:40:00  0460              BATTERY   
2503110 2006-05-10 19:40:00  0460              BATTERY   
3242556 2007-11-16 00:01:00  1150   DECEPTIVE PRACTICE   
3242583 2007-01-07 05:00:00  1140   DECEPTIVE PRACTICE   
3636357 2005-11-10 10:10:00  2825        OTHER OFFENSE   
4031835 2008-07-18 20:00:00  1310      CRIMINAL DAMAGE   
4058378 2008-07-13 18:08:55  1811            NARCOTICS   
4077371 2008-08-23 21:23:00  1563          SEX OFFENSE   
4082463 2008-08-26 23:40:00  0320              ROBBERY   
4084440 2008-08-23 12:00:00  1122   DECEPTIVE PRACTICE   
4098626 2008-09-07 13:00:00  4387        OTHER OFFENSE   
4099705 2008-09-09 17:00:00  0910  MOTOR VEHICLE THEFT   
4115301 2008-09-21 22:00:00  0550              ASSAULT   
4116498 2008-09-22 21:10:00  0560              ASSAULT   
4118831 2008-09-20 20:00:00  1563          SEX OFFENSE   
4128199 2008-09-03 12:00:00  0910  MOTOR VEHICLE THEFT   
4128562 2008-10-03 12:50:00  1130   DECEPTIVE PRACTICE   
4138695 2008-10-11 15:00:00  0910  MOTOR VEHICLE THEFT   
4142813 2008-10-15 16:10:00  0610             BURGLARY   
4145693 2008-10-16 23:00:00  0313              ROBBERY   
4146444 2008-10-18 06:30:00  0291  CRIM SEXUAL ASSAULT   
4146903 2008-10-17 22:39:00  0320              ROBBERY   
4151617 2008-10-22 08:30:00  0610             BURGLARY   
4155882 2008-08-21 14:30:00  1152   DECEPTIVE PRACTICE   
4156879 2008-10-17 12:00:00  0843                THEFT   
4157285 2008-10-27 21:19:00  0920  MOTOR VEHICLE THEFT   
4157931 2008-10-25 17:45:00  0330              ROBBERY   
4158171 2008-10-28 12:30:00  0890                THEFT   
...                     ...   ...                  ...   
5290308 2008-10-15 16:10:00  0610             BURGLARY   
5293188 2008-10-16 23:00:00  0313              ROBBERY   
5293939 2008-10-18 06:30:00  0291  CRIM SEXUAL ASSAULT   
5294398 2008-10-17 22:39:00  0320              ROBBERY   
5299112 2008-10-22 08:30:00  0610             BURGLARY   
5303377 2008-08-21 14:30:00  1152   DECEPTIVE PRACTICE   
5304374 2008-10-17 12:00:00  0843                THEFT   
5304780 2008-10-27 21:19:00  0920  MOTOR VEHICLE THEFT   
5305426 2008-10-25 17:45:00  0330              ROBBERY   
5305666 2008-10-28 12:30:00  0890                THEFT   
5305759 2008-10-21 20:15:00  1506         PROSTITUTION   
5305939 2008-10-24 17:45:00  1345      CRIMINAL DAMAGE   
5305946 2008-10-27 19:14:20  0560              ASSAULT   
5306024 2008-10-28 19:10:00  2900    WEAPONS VIOLATION   
5306298 2008-10-26 03:02:00  0486              BATTERY   
5307407 2008-10-29 05:09:00  0430              BATTERY   
5307773 2008-10-14 21:44:00  143A    WEAPONS VIOLATION   
5308930 2008-10-10 23:57:00  0554              ASSAULT   
5310726 2008-10-31 18:00:00  1310      CRIMINAL DAMAGE   
5310915 2008-11-01 16:10:00  0840                THEFT   
5311440 2008-10-30 15:25:00  1110   DECEPTIVE PRACTICE   
5312370 2008-11-02 20:15:00  0330              ROBBERY   
5313341 2008-11-03 17:09:00  0910  MOTOR VEHICLE THEFT   
5314205 2008-10-21 22:00:00  0890                THEFT   
5314944 2008-11-04 20:14:29  1320      CRIMINAL DAMAGE   
5315320 2008-11-01 23:00:00  0460              BATTERY   
5316080 2008-10-29 13:50:56  0460              BATTERY   
5838285 2010-03-26 20:00:00  0890                THEFT   
6455332 2011-12-07 09:00:00  0820                THEFT   
7022513 2013-06-07 20:13:00  1811            NARCOTICS   

                              Description               Location Description  \
1802297                 CREDIT CARD FRAUD                         RESTAURANT   
1817066                    FORCIBLE ENTRY                  CONSTRUCTION SITE   
1969394                            SIMPLE                              OTHER   
2503110                            SIMPLE                              OTHER   
3242556                 CR

The number of missing 'District' observations: 

In [72]:
print(len(nullDistrict))

91


In [95]:
dataset['District'] = dataset['District'].fillna('UNKNOWN')

In [96]:
dataset['District'].isnull().sum()

0

In [76]:
locDistrict = dataset.loc[(dataset['Latitude'].isnull() == True) & (dataset['Longitude'].isnull() == True)]
locDistrict

Date  IUCR                Primary Type  \
0       2004-01-01 00:01:00  0840                       THEFT   
2       2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3       2004-12-30 20:00:00  0840                       THEFT   
5       2004-08-01 00:01:00  0841                       THEFT   
8       2004-09-16 10:00:00  0890                       THEFT   
11      2004-11-01 00:01:00  0840                       THEFT   
13      2004-11-01 00:00:00  0840                       THEFT   
19      2002-06-14 16:00:00  0266         CRIM SEXUAL ASSAULT   
21      2004-01-01 00:00:00  0840                       THEFT   
27      2004-12-13 13:00:00  1130          DECEPTIVE PRACTICE   
28      2004-10-06 09:00:00  0841                       THEFT   
29      2004-01-12 15:00:00  2825               OTHER OFFENSE   
30      2004-09-17 09:00:00  1130          DECEPTIVE PRACTICE   
38      2004-01-01 00:01:00  0843                       THEFT   
41      2004-01-01 00:01:00  1754  OFFENSE INVOLVING CHILDREN   
42      2002-07-01 00:00:00  0840                       THEFT   
43      2004-11-01 00:01:00  1562                 SEX OFFENSE   
44      2004-03-01 09:00:00  1120          DECEPTIVE PRACTICE   
45      2004-06-29 00:01:00  1120          DECEPTIVE PRACTICE   
46      2004-06-27 15:30:00  0460                     BATTERY   
48      2004-07-08 00:01:00  0263         CRIM SEXUAL ASSAULT   
49      2004-01-01 00:01:00  0840                       THEFT   
51      2004-08-01 12:00:00  0841                       THEFT   
55      2004-07-03 16:30:00  0610                    BURGLARY   
56      2004-10-13 12:00:00  0840                       THEFT   
57      2004-06-15 12:00:00  2820               OTHER OFFENSE   
58      2004-11-01 00:01:00  0840                       THEFT   
59      2004-01-01 00:00:00  1752  OFFENSE INVOLVING CHILDREN   
60      2004-02-09 14:00:00  0840                       THEFT   
61      2004-07-09 00:01:00  0840                       THEFT   
...                     ...   ...                         ...   
7941191 2016-07-18 00:00:00  0266         CRIM SEXUAL ASSAULT   
7941192 2017-01-15 09:00:00  1153          DECEPTIVE PRACTICE   
7941193 2013-11-20 09:00:00  1153          DECEPTIVE PRACTICE   
7941194 2017-01-16 10:00:00  0610                    BURGLARY   
7941195 2016-12-15 10:00:00  2851      PUBLIC PEACE VIOLATION   
7941196 2017-01-15 23:59:00  0820                       THEFT   
7941197 2017-01-14 06:05:00  0560                     ASSAULT   
7941198 2016-12-19 19:03:00  1206          DECEPTIVE PRACTICE   
7941199 2016-10-10 17:00:00  1210          DECEPTIVE PRACTICE   
7941200 2016-06-07 09:00:00  1153          DECEPTIVE PRACTICE   
7941201 2016-11-11 01:30:00  0930         MOTOR VEHICLE THEFT   
7941202 2017-01-06 22:00:00  0281         CRIM SEXUAL ASSAULT   
7941203 2017-01-01 00:00:00  0810                       THEFT   
7941204 2016-11-16 16:00:00  0620                    BURGLARY   
7941205 2017-01-13 08:32:00  2825               OTHER OFFENSE   
7941206 2016-04-25 09:00:00  4387               OTHER OFFENSE   
7941207 2016-12-20 18:00:00  1153          DECEPTIVE PRACTICE   
7941208 2017-01-15 12:00:00  1150          DECEPTIVE PRACTICE   
7941209 2016-12-02 00:01:00  1153          DECEPTIVE PRACTICE   
7941210 2016-12-14 09:28:00  1206          DECEPTIVE PRACTICE   
7941211 2017-01-11 13:00:00  5093                NON-CRIMINAL   
7941212 2016-12-17 08:00:00  1130          DECEPTIVE PRACTICE   
7941213 2016-04-17 07:00:00  1150          DECEPTIVE PRACTICE   
7941214 2017-01-16 14:30:00  0820                       THEFT   
7941215 2017-01-08 11:00:00  0890                       THEFT   
7941216 2017-01-16 16:00:00  1150          DECEPTIVE PRACTICE   
7941217 2016-12-20 08:00:00  1153          DECEPTIVE PRACTICE   
7941218 2017-01-15 00:00:00  1153          DECEPTIVE PRACTICE   
7941219 2017-01-18 12:00:00  1152          DECEPTIVE PRACTICE   
7941223 2017-01-02 09:14:00  0110                    HOMIC

Number of observations where both Latitude and Longitude are missing: 

In [77]:
len(locDistrict)

105573

Number of observations where Latitude info is missing:

In [78]:
 len(dataset.loc[(dataset['Latitude'].isnull() == True)])

105573

Number of observations where Longitude info is missing:

In [79]:
 len(dataset.loc[(dataset['Longitude'].isnull() == True)])

105574

In [105]:
dataset['Longitude'] = dataset['Longitude'].fillna(0.0)

In [109]:
dataset['Longitude'].isnull().sum()

0

In [110]:
dataset['Latitude'] = dataset['Latitude'].fillna(0.0)

In [111]:
dataset['Latitude'].isnull().sum()

0


## Unnesting Information on Crime Hour, Day, Weekday, and Month using Dates
* Hour is what hour in the day did the crime occur
* Day is what day in the month (1st day of the month, 28th day of the month..) did ht observation happen
* Weekday is what day in the week did the crime occur (Monday =0, Tuesday =1, Wednesday =2, ..., Sunday = 6) 
* Month is what month it happend 

In [117]:
dataset = dataset.assign(Month = dataset["Date"].dt.month)

In [119]:
dataset.head()

Date  IUCR                Primary Type  \
0 2004-01-01 00:01:00  0840                       THEFT   
1 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3 2004-12-30 20:00:00  0840                       THEFT   
4 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

  District    Year Latitude  Longitude  Month  
0        4  2004.0        0   0.000000      1  
1        9  2003.0  41.8172 -87.637328      3  
2       14  2004.0        0   0.000000      6  
3       25  2004.0        0   0.000000     12  
4       22  2003.0  41.6918 -87.635116      5

In [120]:
dataset = dataset.assign(Day = dataset["Date"].dt.day)

In [121]:
dataset.head()

Date  IUCR                Primary Type  \
0 2004-01-01 00:01:00  0840                       THEFT   
1 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3 2004-12-30 20:00:00  0840                       THEFT   
4 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

  District    Year Latitude  Longitude  Month  Day  
0        4  2004.0        0   0.000000      1    1  
1        9  2003.0  41.8172 -87.637328      3    1  
2       14  2004.0        0   0.000000      6   20  
3       25  2004.0        0   0.000000     12   30  
4       22  2003.0  41.6918 -87.635116      5    1

In [122]:
dataset = dataset.assign(Hour = dataset["Date"].dt.hour)

In [123]:
dataset.head()

Date  IUCR                Primary Type  \
0 2004-01-01 00:01:00  0840                       THEFT   
1 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3 2004-12-30 20:00:00  0840                       THEFT   
4 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

  District    Year Latitude  Longitude  Month  Day  Hour  
0        4  2004.0        0   0.000000      1    1     0  
1        9  2003.0  41.8172 -87.637328      3    1     0  
2       14  2004.0        0   0.000000      6   20    11  
3       25  2004.0        0   0.000000     12   30    20  
4       22  2003.0  41.6918 -87.635116      5    1     1

In [126]:
dataset = dataset.assign(Weekday = dataset["Date"].dt.weekday)

In [127]:
dataset.head()

Date  IUCR                Primary Type  \
0 2004-01-01 00:01:00  0840                       THEFT   
1 2003-03-01 00:00:00  2825               OTHER OFFENSE   
2 2004-06-20 11:00:00  1752  OFFENSE INVOLVING CHILDREN   
3 2004-12-30 20:00:00  0840                       THEFT   
4 2003-05-01 01:00:00  0841                       THEFT   

                      Description Location Description  Arrest  Domestic  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False     False   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False      True   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False     False   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False     False   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False     False   

  District    Year Latitude  Longitude  Month  Day  Hour  Weekday  
0        4  2004.0        0   0.000000      1    1     0        3  
1        9  2003.0  41.8172 -87.637328      3    1     0        5  
2       14  2004.0        0   0.000000      6   20    11        6  
3       25  2004.0        0   0.000000     12   30    20        3  
4       22  2003.0  41.6918 -87.635116      5    1     1        3

In [128]:
dataset = dataset.drop(['Date'], axis  = 1)

In [129]:
dataset.head()

IUCR                Primary Type                     Description  \
0  0840                       THEFT   FINANCIAL ID THEFT: OVER $300   
1  2825               OTHER OFFENSE         HARASSMENT BY TELEPHONE   
2  1752  OFFENSE INVOLVING CHILDREN   AGG CRIM SEX ABUSE FAM MEMBER   
3  0840                       THEFT   FINANCIAL ID THEFT: OVER $300   
4  0841                       THEFT  FINANCIAL ID THEFT:$300 &UNDER   

  Location Description  Arrest  Domestic District    Year Latitude  Longitude  \
0            RESIDENCE   False     False        4  2004.0        0   0.000000   
1            RESIDENCE   False      True        9  2003.0  41.8172 -87.637328   
2            RESIDENCE   False     False       14  2004.0        0   0.000000   
3                OTHER   False     False       25  2004.0        0   0.000000   
4            RESIDENCE   False     False       22  2003.0  41.6918 -87.635116   

   Month  Day  Hour  Weekday  
0      1    1     0        3  
1      3    1     0        5  
2      6   20    11        6  
3     12   30    20        3  
4      5    1     1        3

## Export New Dataset with No Missing Values 

This will export the dataset to the crimesInChicagoData folder as "dataset.csv"

In [130]:
dataset = dataset.sort_index()

In [131]:
dataset.to_csv("../../crimesInChicagoData/dataset.csv")